feat(mobile): optimized thumbnail rendering

mertalev opened the pull request:
## Description

Uses custom optimized image widgets, a better thumbnail preparation process and request cancellation for offscreen images to make thumbnails render significantly more quickly than before, and in significantly higher quality.

Later PRs can improve this further by improving colors with Display P3 and such.

Notably, I also tested a platform view-based implementation based on UIImage. However, the high number of platform views made scrolling stutter-y and used more memory, so transferring the image to Flutter's rendering engine is more efficient. I want to try this approach for the asset viewer in a later PR though.

Specific improvements:
- Thumbnails are higher resolution and quality
- Decoding and rendering now use less memory and CPU
- On Android, thumbnails are now leveraging pre-generated thumbnails from the OS
- No longer need a disk cache for local thumbnails, lowering storage usage
- Removed redundant platform-side disk and memory caching that was done in addition to Flutter-side caching
- Multi-threaded execution in the background, never blocking the UI thread (except thumbhashes for now)
- Higher quality thumbnails mean they can double as the asset viewer placeholder, making them more consistently available immediately
- Asset viewer now uses higher resolution images
- Remote assets now consistently show a thumbhash rather than a gray box when loading
- Scrubbing now shows thumbhashes
- Scrolling shows the thumbnail instantly much more reliably
- Improved responsiveness by cancelling unnecessary thumbnail requests

To do:
- [x] Android implementation
- [x] Support contain vs cover fit
- [x] Code cleanup
- [ ] Test pages and in different conditions (fast vs slow device, fast vs slow internet, etc.)

## How Has This Been Tested?

Tested by scrolling and scrubbing in the timeline.

mertalev committed:
thumbhash improvements

thumbhash render box

refactor

wip

rebase

shenlong-tanwen commented on mobile/lib/presentation/widgets/images/thumbnail.widget.dart:
```diff
@@ -58,11 +57,7 @@ OctoPlaceholderBuilder _blurHashPlaceholderBuilder(
 }) {
   return (context) => thumbHash == null
       ? const ThumbnailPlaceholder()
-      : FadeInPlaceholderImage(
-          placeholder: const ThumbnailPlaceholder(),
-          image: ThumbHashProvider(thumbHash: thumbHash),
-          fit: fit ?? BoxFit.cover,
-        );
```
The reason we did it this way was to prevent the flashing when the widget changes from the placeholder to the thumbhash. Can you try wrapping them around with `AnimatedSwitcher` to make the transition a bit smooth?

mertalev replied:
The thumbhash is the placeholder, I don't see why you'd want to delay displaying it. You want it to show as quickly as possible because the thumbnail is going to display soon. Also the less complex rendering stuff is in here, the more likely the thumbhashes get to load in the first frame.

mertalev replied:
Did you test with a release build? I generally see the thumbhash straight away even when scrolling quickly. I haven't seen a blank screen with indicators like you describe. Also, I usually don't get to see the thumbhash on main because it's too slow to come in, so it's more like placeholder -> thumbnail.

alextran1502 commented:
Hi Mert, do you have any changes that you want to push for testing?

mertalev commented:
> Hi Mert, do you have any changes that you want to push for testing?

I pushed my thumbnail optimizations so you can give it a try. It's very much WIP (only iOS is implemented), but the results should speak for themselves.

mertalev committed:
kotlin impl, avoid message passing overhead

mertalev committed:
draw to buffer

inline

scale video frame when possible

account for different dimensions

mertalev committed:
light at the end of the tunnel

mertalev committed:
transition

mertalev committed:
cleanup

restore xcode settings

formatting

restore xcode settings

fix rebase

alextran1502 commented on mobile/ios/Runner/Runner.entitlements:
```diff
@@ -9,8 +9,6 @@
 	<key>com.apple.developer.networking.wifi-info</key>
 	<true/>
 	<key>com.apple.security.application-groups</key>
-	<array>
```
Should this fine be reverted?

alextran1502 commented on mobile/ios/Runner/RunnerProfile.entitlements:
```diff
@@ -11,8 +11,6 @@
 	<key>com.apple.developer.networking.wifi-info</key>
 	<true/>
 	<key>com.apple.security.application-groups</key>
```
Should this fine be reverted?

alextran1502 commented on mobile/lib/presentation/widgets/asset_viewer/asset_viewer.page.dart:
```diff
@@ -121,11 +121,11 @@ class _AssetViewerState extends ConsumerState<AssetViewer> {
 
   bool get showingBottomSheet =>
-      ref.read(assetViewerProvider.select((s) => s.showingBottomSheet));
+      ref.read(assetViewerProvider).showingBottomSheet;
```
I believe this will trigger rerendering when any state changes in the assetViewerProvider

mertalev replied:
It's `ref.read`, so it doesn't trigger a state change regardless

mertalev committed:
disable thumbhash for local assets

better update condition

load from provider first

show thumbhash for merged assets when scrubbing

fix disposal

mertalev committed:
formatting

mertalev committed:
preserve subtree

not growable

minor cleanup

mertalev committed:
buttery hero animations

buttery hero animation for remote assets

mertalev committed:
add ffi dependency

formatting

mertalev committed:
request cancellation

mertalev committed:
update constants

mertalev committed:
remove unused parameter

mertalev committed:
fix ref cycle

mertalev committed:
ios fixes and improvements

mertalev commented:
Closing in favor of separate PRs.
